### PR TITLE
Login

### PR DESCRIPTION
**Pull request by VyacheslavBukach**

### CODE REVIEW

File: lib/blocs/auth_bloc/auth_state.dart [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+abstract class AuthState {}
+
+class Loading extends AuthState {}
+
+class Authenticated extends AuthState {}
+
+class UnAuthenticated extends AuthState {}
+
+class AuthError extends AuthState {
+  AuthError(this.error);
```
Comment (MaximSemashko) [inline lib/blocs/auth_bloc/auth_state.dart]: firstly should be fields after that constructors

File: lib/data/repositories/firebase_auth_repository_impl.dart [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+class FirebaseAuthRepositoryImpl implements AuthRepository {
+  final FirebaseAuth _firebaseAuth = FirebaseAuth.instance;
+
+  @override
+  Future<void> signIn(String email, String password) async {
+    try {
+      await _firebaseAuth.signInWithEmailAndPassword(
+          email: email, password: password);
+    } on FirebaseAuthException catch (e) {
+      if (e.code == 'user-not-found') {
```
Comment (MaximSemashko) [inline lib/data/repositories/firebase_auth_repository_impl.dart]: all error codes for example 'user-not-found' should be extract to constants
`const  _kUserNotFound = 'user-not-found'`

File: lib/domain/entities/book.dart [hunk 3]
```diff
@@ -0,0 +1,22 @@
+class Book {
+  Book({
+    required this.uuid,
+    required this.title,
+    required this.author,
+  });
+
+  String uuid;
```
Comment (MaximSemashko) [inline lib/domain/entities/book.dart]: fields should be before constructor

File: lib/domain/usecases/auth/auth_usecase.dart [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import 'package:my_books/domain/repositories/auth_repository.dart';
+
+class AuthUseCase {
+  AuthUseCase({required this.authRepo});
+
+  final AuthRepository authRepo;
+
+  Future login(String email, String password) async {
```
Comment (MaximSemashko) [inline lib/domain/usecases/auth/auth_usecase.dart]: can be changed to `Future<void>`

File: lib/domain/usecases/auth/auth_usecase.dart [hunk 5]
```diff
@@ -0,0 +1,19 @@
+import 'package:my_books/domain/repositories/auth_repository.dart';
+
+class AuthUseCase {
```
Comment (MaximSemashko) [inline lib/domain/usecases/auth/auth_usecase.dart]: can be splitted to several use cases
for instance LoginUseCase, RegisterUseCase etc

File: lib/main.dart [hunk 6]
```diff
@@ -1,115 +1,51 @@
...
+                RepositoryProvider.of<FirebaseAuthRepositoryImpl>(context),
+          ),
+        ),
+        child: MaterialApp(
+          onGenerateTitle: (BuildContext context) =>
+              AppLocalizations.of(context)?.app_name ?? '',
+          localizationsDelegates: AppLocalizations.localizationsDelegates,
+          supportedLocales: AppLocalizations.supportedLocales,
+          home: StreamBuilder<User?>(
+            stream: FirebaseAuth.instance.authStateChanges(),
```
Comment (MaximSemashko) [inline lib/main.dart]: should be extracted to repository

File: lib/presentation/ui/books_screen.dart [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+import '../../blocs/auth_bloc/auth_bloc.dart';
+
+class BooksScreen extends StatelessWidget {
+  const BooksScreen({Key? key}) : super(key: key);
+
+  @override
+  Widget build(BuildContext context) {
+    final user = FirebaseAuth.instance.currentUser;
+    return Scaffold(
```
Comment (MaximSemashko) [inline lib/presentation/ui/books_screen.dart]: add empty line before return

File: lib/presentation/ui/books_screen.dart [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+import 'package:my_books/presentation/ui/sign_in_screen.dart';
+
+import '../../blocs/auth_bloc/auth_bloc.dart';
+
+class BooksScreen extends StatelessWidget {
+  const BooksScreen({Key? key}) : super(key: key);
+
+  @override
+  Widget build(BuildContext context) {
+    final user = FirebaseAuth.instance.currentUser;
```
Comment (MaximSemashko) [inline lib/presentation/ui/books_screen.dart]: Should be extracted to the repository

File: lib/presentation/ui/books_screen.dart [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+              (route) => false,
+            );
+          }
+        },
+        child: Center(
+          child: Column(
+            mainAxisAlignment: MainAxisAlignment.center,
+            children: [
+              Text(
+                'Email: \n ${user?.email}',
```
Comment (MaximSemashko) [inline lib/presentation/ui/books_screen.dart]: Extract to the Locale

File: lib/presentation/ui/books_screen.dart [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+                style: const TextStyle(fontSize: 24),
+                textAlign: TextAlign.center,
+              ),
+              const SizedBox(height: 16),
+              ElevatedButton(
+                child: const Text('Sign Out'),
```
Comment (MaximSemashko) [inline lib/presentation/ui/books_screen.dart]: The same

File: lib/blocs/auth_bloc/auth_state.dart (hunk 1 above)
Comment (VyacheslavBukach) [reply lib/blocs/auth_bloc/auth_state.dart]: done

File: lib/data/repositories/firebase_auth_repository_impl.dart (hunk 2 above)
Comment (VyacheslavBukach) [reply lib/data/repositories/firebase_auth_repository_impl.dart]: done

File: lib/domain/entities/book.dart (hunk 3 above)
Comment (VyacheslavBukach) [reply lib/domain/entities/book.dart]: done

File: lib/domain/usecases/auth/auth_usecase.dart (hunk 4 above)
Comment (VyacheslavBukach) [reply lib/domain/usecases/auth/auth_usecase.dart]: done

File: lib/domain/usecases/auth/auth_usecase.dart (hunk 5 above)
Comment (VyacheslavBukach) [reply lib/domain/usecases/auth/auth_usecase.dart]: done

File: lib/presentation/ui/books_screen.dart (hunk 7 above)
Comment (VyacheslavBukach) [reply lib/presentation/ui/books_screen.dart]: done

File: lib/main.dart (hunk 6 above)
Comment (VyacheslavBukach) [reply lib/main.dart]: done

File: lib/presentation/ui/books_screen.dart (hunk 9 above)
Comment (VyacheslavBukach) [reply lib/presentation/ui/books_screen.dart]: done

File: lib/presentation/ui/books_screen.dart (hunk 10 above)
Comment (VyacheslavBukach) [reply lib/presentation/ui/books_screen.dart]: done

File: lib/presentation/ui/books_screen.dart (hunk 8 above)
Comment (VyacheslavBukach) [reply lib/presentation/ui/books_screen.dart]: done

**Pull request merged**